scrapy/utils/item.py

```diff
@@ -0,0 +1,104 @@
+
+
+def is_item_like(obj):
+    """
+    Returns True if *obj* is considered a Scrapy *item*, False otherwise.
+
+    An object is considered an *item* if it is:
+    - a scrapy.item.BaseItem or dict instance (or any subclass)
+    - a dataclass object
+    """
+    return isinstance(obj, (BaseItem, dict)) or _is_dataclass_instance(obj)
+
+
+class ItemAdapter:
+    """
+    Wrapper class to interact with items. It provides a common interface for components
+    such as middlewares and pipelines to extract and set data without having to take
+    the item's implementation (scrapy.Item, dict, dataclass) into account.
+    """
+
+    def __init__(self, item):
+        if not is_item_like(item):
+            raise TypeError("Expected a valid item, got %r instead: %s" % (type(item), item))
+        self.item = item
+
+    def __repr__(self):
+        return "ItemAdapter for type %s: %r" % (self.item.__class__.__name__, self.item)
+
+    def __contains__(self, field_name):
+        if _is_dataclass_instance(self.item):
+            from dataclasses import fields
+            return field_name in (f.name for f in fields(self.item))
+        return field_name in self.item
+
+    def __getitem__(self, field_name):
+        if _is_dataclass_instance(self.item):
+            if field_name in self:
+                return getattr(self.item, field_name)
+            raise KeyError(field_name)
+        return self.item[field_name]
+
+    def __setitem__(self, field_name, value):
+        if _is_dataclass_instance(self.item):
+            if field_name in self:
+                setattr(self.item, field_name, value)
+            else:
+                raise KeyError(
+                    "%s does not support field: %s" % (self.item.__class__.__name__, field_name))
+        else:
+            self.item[field_name] = value
+
+    def get(self, field_name, default=None):
+        if _is_dataclass_instance(self.item):
+            return getattr(self.item, field_name, default)
+        return self.item.get(field_name, default)
+
+    def get_field(self, field_name):
+        """
+        Returns the appropriate class:`scrapy.item.Field` object if the wrapped item
+        is a BaseItem object, None otherwise.
+        """
+        if isinstance(self.item, BaseItem):
+            return self.item.fields.get(field_name)
+        return None
+
+    def asdict(self):
+        """
+        Return a class:`dict` instance with the same data as the wrapped item.
+        Returns a shallow copy of the wrapped item if it is already a dict object.
+        """
+        if _is_dataclass_instance(self.item):
+            from dataclasses import asdict
+            return asdict(self.item)
+        elif isinstance(self.item, dict):
+            return self.item.copy()
+        elif isinstance(self.item, BaseItem):
+            return dict(self.item)
+
+    def field_names(self):
+        """
+        Returns a generator with the names of the item's fields
```

Comment:
I'd prefer keys method, as key view is more useful than a generator. For example, for dict.keys() `__repr__` shows actual keys. List is also ok. I can't imagine a case where having all item keys in memory (as opposed to making them on-by-one in  a generator) would cause memory issues.

Reply:
Makes sense, I'll make it return a list to keep the return type consistent (IIRC there is no way to create `dict_keys` objects in user code).
Regarding the name, I think it'd be better to keep it as it is now, to make it clear that it returns *all* field names. This method is only used in the `scrapy.exporters.BaseItemExporter` class, where sometimes there is the need to know all the defined fields for a given item. In `BaseItem`-based items, `Ìtem.keys()` and `Item.fields.keys()` can be different if a field is defined but unset:

```python
In [1]: from scrapy.item import Item, Field
   ...: class TestItem(Item):
   ...:     name = Field()
   ...:     value = Field()
   ...: i = TestItem(name="asdf")
   ...: print(i.keys())
   ...: print(i.fields.keys())
   ...:
dict_keys(['name'])
dict_keys(['name', 'value'])
```

Reply:
I _think_ you should be getting `.keys()` method for free if you implement MutableMapping, but I haven't actually checked it.

Reply:
Nice, indeed:
```python
In [1]: from scrapy.utils.item import ItemAdapter                                                                                                                                                                                             

In [2]: ItemAdapter({}).keys()                                                                                                                                                                                                                
Out[2]: KeysView(ItemAdapter for type dict: {})
```
I think the `field_names` function might still be necessary though, since the current exporter implementation needs all fields when exporting empty fields, and `keys` only returns those which are set:
```python
In [10]: from scrapy.item import Item, Field 
    ...: from scrapy.utils.item import ItemAdapter 
    ...: class TestItem(Item): 
    ...:     name = Field() 
    ...:     value = Field() 
    ...: i = TestItem(name="asdf") 
    ...: a = ItemAdapter(i) 
    ...: print(i.keys()) 
    ...: print(i.fields.keys()) 
    ...: print(list(a.keys()))                                                                                                                                                                                                                
dict_keys(['name'])
dict_keys(['name', 'value'])
['name']
```